neon/src/connection.js

```diff
@@ -0,0 +1,1046 @@
+    while (sent.length)
+      queryError(sent.shift(), err)
+  }
+
+  function errored(err) {
+    stream && (stream.destroy(err), stream = null)
+    query && queryError(query, err)
+    initial && (queryError(initial, err), initial = null)
+  }
+
+  function queryError(query, err) {
+    'query' in err || 'parameters' in err || Object.defineProperties(err, {
+      stack: { value: err.stack + query.origin.replace(/.*\n/, '\n'), enumerable: options.debug },
+      query: { value: query.string, enumerable: options.debug },
+      parameters: { value: query.parameters, enumerable: options.debug },
+      args: { value: query.args, enumerable: options.debug },
+      types: { value: query.statement && query.statement.types, enumerable: options.debug }
+    })
+    query.reject(err)
+  }
+
+  function end() {
+    return ending || (
+      !connection.reserved && onend(connection),
+      !connection.reserved && !initial && !query && sent.length === 0
+        ? (terminate(), new Promise(r => socket && socket.readyState !== 'closed' ? socket.once('close', r) : r()))
+        : ending = new Promise(r => ended = r)
+    )
+  }
+
+  function terminate() {
+    terminated = true
+    if (stream || query || initial || sent.length)
+      error(Errors.connection('CONNECTION_DESTROYED', options))
+
+    clearImmediate(nextWriteTimer)
+    if (socket) {
+      socket.removeListener('data', data)
+      socket.removeListener('connect', connected)
+      socket.readyState === 'open' && socket.end(b().X().end())
+    }
+    ended && (ended(), ending = ended = null)
+  }
+
+  async function closed(hadError) {
+    incoming = Buffer.alloc(0)
+    remaining = 0
+    incomings = null
+    clearImmediate(nextWriteTimer)
+    socket.removeListener('data', data)
+    socket.removeListener('connect', connected)
+    idleTimer.cancel()
+    lifeTimer.cancel()
+    connectTimer.cancel()
+
+    socket.removeAllListeners()
+    socket = null
+
+    if (initial)
+      return reconnect()
+
+    !hadError && (query || sent.length) && error(Errors.connection('CONNECTION_CLOSED', options, socket))
+    closedDate = performance.now()
+    hadError && options.shared.retries++
+    delay = (typeof backoff === 'function' ? backoff(options.shared.retries) : backoff) * 1000
+    onclose(connection, Errors.connection('CONNECTION_CLOSED', options, socket))
+  }
+
+  /* Handlers */
+  function handle(xs, x = xs[0]) {
+    (
+      x === 68 ? DataRow :                   // D
+      x === 100 ? CopyData :                 // d
+      x === 65 ? NotificationResponse :      // A
+      x === 83 ? ParameterStatus :           // S
+      x === 90 ? ReadyForQuery :             // Z
+      x === 67 ? CommandComplete :           // C
+      x === 50 ? BindComplete :              // 2
+      x === 49 ? ParseComplete :             // 1
+      x === 116 ? ParameterDescription :     // t
+      x === 84 ? RowDescription :            // T
+      x === 82 ? noop :                      // R
```

Comment:
We can safely ignore the "Authentication" request because we've already sent the cleartext password.